sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuPythonUDF.scala

```diff
@@ -75,16 +76,58 @@ case class GpuPythonUDF(
 
   override def nullable: Boolean = true
 
-  override lazy val canonicalized: Expression = {
-    val canonicalizedChildren = children.map(_.canonicalized)
-    // `resultId` can be seen as cosmetic variation in PythonUDF, as it doesn't affect the result.
-    this.copy(resultId = ExprId(-1)).withNewChildren(canonicalizedChildren)
-  }
-
   // Support window things
   override val windowInputProjection: Seq[Expression] = Seq.empty
+
   override def windowAggregation(
       inputs: Seq[(ColumnVector, Int)]): RollingAggregationOnColumn = {
     throw new UnsupportedOperationException(s"GpuPythonUDF should run in a Python process.")
   }
 }
+
+case class GpuPythonUDF(
+    name: String,
+    override val func: PythonFunction,
+    dataType: DataType,
+    children: Seq[Expression],
+    evalType: Int,
+    udfDeterministic: Boolean,
+    resultId: ExprId = NamedExpression.newExprId)
+  extends GpuPythonFunction(name, func, dataType, children, evalType, udfDeterministic, resultId) {
+  override lazy val canonicalized: Expression = {
+    val canonicalizedChildren = children.map(_.canonicalized)
+    // `resultId` can be seen as cosmetic variation in PythonUDF, as it doesn't affect the result.
+    this.copy(resultId = ExprId(-1)).withNewChildren(canonicalizedChildren)
+  }
+}
+
+case class GpuPythonUDAF(
+    name: String,
+    override val func: PythonFunction,
+    dataType: DataType,
+    children: Seq[Expression],
+    evalType: Int,
+    udfDeterministic: Boolean,
+    resultId: ExprId = NamedExpression.newExprId)
+  extends GpuPythonFunction(name, func, dataType, children, evalType, udfDeterministic, resultId)
+    with GpuAggregateFunction {
+  override lazy val canonicalized: Expression = {
+    val canonicalizedChildren = children.map(_.canonicalized)
+    // `resultId` can be seen as cosmetic variation in PythonUDF, as it doesn't affect the result.
+    this.copy(resultId = ExprId(-1)).withNewChildren(canonicalizedChildren)
+  }
+
+  override val initialValues: Seq[Expression] = Seq.empty[Expression]
+
+  override val inputProjection: Seq[Expression] = Seq.empty[Expression]
+
+  override val updateAggregates: Seq[CudfAggregate] = Seq.empty[CudfAggregate]
+
+  override val mergeAggregates: Seq[CudfAggregate] = Seq.empty[CudfAggregate]
+
+  override val evaluateExpression: Expression = null
+
+  override def aggBufferAttributes: Seq[AttributeReference] = {
+    throw new IllegalStateException()
```

Comment:
It would be nice to have a more descriptive error message here.

Reply:
I am not sure what descriptive message to throw here. 
Will it make sense if we instead throw an `UnsupportedOperationException("GpyPythonUDAF isn't a real aggregate function)`

Reply:
That works, as does following what Spark does for PythonUDAF for the same function (see UnevaluableAggregateFunc).

That makes me think -- why are we returning empty sequences for some of these functions but not others -- should any of these aggregate-related functions be legitimately called or should they all be throwing an error?  I'd rather the code throw errors if it's not supposed to be called rather than return an empty sequence.

Reply:
Yes, I think it would make sense for all of the aggregate-related functions to throw the same exception

Reply:
The problem is that most of these are `val` not methods.

Reply:
Ah right, for vals we won't be able to do this and empty sequence is about the best we can do.